### DISCUSSION TITLE
Disabling endpoint aggregation

### DISCUSSION THREAD

**Discussion by WojTecH94** [DefectDojo/django-DefectDojo#9669] [Q&A]
Hello,

Is it possible to disable endpoints aggregation for the same vulnerability?
I mean if you import scan and there are multiple hosts with the same vulnerability DefectDojo merges them to one vulnerability entry.
It would be useful to disable this behavior. For example: you want to make report of vulnerabilities for single host and send it to host's admin but in the report he can see that there are the same vulnerabilities on different hosts which might be treated as information disclosure in some cases...

**Comment by mtesauro** [comment]
Have you tried to look from the Endpoints / Vulnerable endpoints perspective? This should provide reporting/filtering based on endpoints instead of finding.
![image](https://github.com/DefectDojo/django-DefectDojo/assets/3422419/61271247-660b-4adc-8817-f26467b12aa3)

**Comment by WojTecH94** [comment]
Yes I have tried, but when you generate report from this view it still contains information about other hosts/endpoints with the same vulnerabilities :(

**Comment by WojTecH94** [comment]
Another case: when you want to accept vulnerability on single host via Full risk acceptance form (not a simple risk acceptance aka bulk edit), you are not able to do this because then vulnerability will be also accepted on other hosts which it exist too and you most probably don't want this.

**Comment by WojTecH94** [comment]
For me the current behavior of Tenable parser is not intentional and I consider it a BUG more and more:
In line https://github.com/DefectDojo/django-DefectDojo/blob/f2c3cf5d44dbd9e8e8739a9814c46e8c75591427/dojo/tools/tenable/csv_format.py#L111
It tries to get host_name field which is most offten not present in tenable csv exports, while few lines later we have variable "host" created in a good way (checking all fields that can contain DNS or ip of the host) which could be used instead:
https://github.com/DefectDojo/django-DefectDojo/blob/f2c3cf5d44dbd9e8e8739a9814c46e8c75591427/dojo/tools/tenable/csv_format.py#L185
When I moved host and port variables definitions before "dupe checker" and used these variables in dupe_key then import acts more naturally for me creating seperate finding for each endpoint.
I need some DD expert to confirm my suspicions or show me some higher logic behind this code :)

**Comment by WojTecH94** [comment]
@FelixHernandez I see you are doing some work on Tenable parser (https://github.com/DefectDojo/django-DefectDojo/pull/9804), maybe you could look at my considerations above?

**Comment by pna-nca** [comment]
Personally, I doubt that endpoint aggregation is a bug. This is a great feature and the way how vulnerability management tool has to work. Otherwise it becomes just another interface to Nessus scan export.

Even if you navigate to Tenable Security Center, you observe the same behavior: endpoints aggregation. But, it will depend on the view. Exactly as one would like to.

In case of DefectDojo, I would suggest to rethink how you organize your vulnerabilities. If you have a project / product, for which you perform scans, a single finding is a finding "for the whole project", whereas it "affects several systems". This is perfectly natural.

In case some of your hosts logically belong to another project, findings there should be considered separately. Thus, just create another project for such hosts.

Reporting problems in DefectDojo indeed exist, but it is not relevant to the "aggregation" discussion.

I would strongly vote against merging the referenced pull request, this will force us to even revert it in our fork as it is the crucial feature.

**Comment by WojTecH94** [comment]
"just create another project for such hosts."
This would lead to basically creating project for each host. There might be hosts within the same project which are more and less exposed and they should not be treated the same.
For me current behavior is not really natural because merging/aggregating endpoints only exist during single scan import. Between different scan imports we have deduplication mechanism which leads to situations where you import another scan with vulnerability (present before) on additional ports and it will not add endpoints to the original(active) finding but it will create new finding and mark it as duplicate(inactive) and you will not see this on the active findings page which is probably the most popular. 
Not to mention that endpoint aggregation even in single scan is not reliable from what I tested:
Sometimes it merges, sometimes creates duplicate and you don't know that there is one more port affected untill you go across duplicates.
What is more... I suspect that when description was added to deduplication fields in tenable hashcode algo (https://github.com/DefectDojo/django-DefectDojo/commit/d0def170741df3257b9aefa2d89997e59fc8a37c) it broke aggregated findings because in tenable you have description created as a concatenation of Nessus Plugin outputs from multiple vulnerabilites right? Now when the description is actually used to generate hash the new findings will have different hash than the aggregated findings with multiple plugin outputs. Also when you have plugin output from different endpoints aggregated into single finding description sometimes you don't know which part belongs to which endpoint and it makes process of fixing vulns harder.

**Comment by WojTecH94** [comment]
All I am trying to say is that current way of presenting "Active findings" not really showing true **ACTIVE** state. When clicking on some active finding we are redirected to "Original" finding which was firstly discovered. During finding lifecycle there could be new scans which discovered the same vulnerability on other endpoints and then duplicates with those new endpoints are created. Meanwhile endpoints in Original finding could have been mitigated, but finding still exists as the duplicates (with endpoints different from original) prevent it from closing. This is really confusing because we still see Original finding with mitigated endpoints and only way to have holistic view is to go through duplicates which is not really logical because they are marked as "inactive"...
If there was a determination to keep the aggregation feature there should be some view other than Original to show the current state of a finding with actual vulnerable/mitigated endpoints (probably using newest duplicate instead of the oldest one). That "newest duplicate" should have status active, while Original should be changed to inactive because it might be outdated in case of endpoints and even if it wouldn't it will be still included in newest duplicate entry. But then what about changes you made to the findings (notes, tags, etc.)? All this seem to be complicated to develop and maintain (for example because duplicates can have different descriptions in Nessus containing dns names, paths etc.).
That is why I think that letting vulnerabilities live "one per endpoint" is the best option. You don't have to worry about things like those mentioned then. 
At least if there was an option in system settings to allow users like me to disable aggregation...
